Add method to run command-line db apps

starbelly opened the pull request:
raise error unless Kernel.system returns true in structure_load()

rafaelfranca commented:
@starbelly seems good. You don't need to open a new PR though. Just force push for the same branch.

simi commented:
@starbelly I think it will be nice to keep this Windows compatible. Can you confirm that `exit_code == 127` works there also?

starbelly commented:
Yeah, definitely should be windows compatible. I'll load up Ruby on a windows box to give this a test, but it looks like it's going to be error code 2 for windows, so we should be able to say `exit_code == 127 || exit_code == 2`. There's probably going to be some additional error checking in windows though (e.g., bad executable, bad permissions, etc. ). 
 I also need to get this working in rails-dev-box first too. Works fine on my own machine, but I was having some oddness with this in the VM and will re-visit that issue now. More to come.

matthewd commented:
I think I'd prefer to just say what we know went wrong: "Failed to execute `psql .. etc`"

Attempting to infer the problem seems far more likely to end in confusion for someone, where the command really is on the path but is itself exiting with 127, or 2, or whatever.

If we just say "we tried to run this, and it failed", firstly, people are more likely to believe any preceding output from the command itself, and then, they'll just try running the command manually to see what's up.

Unless we've actually looked in `ENV['PATH']`, I think it's actively unhelpful for us to imply we have. We've made this mistake before: 660523123103e904622a135940a1c12ad9256629

simi commented:
@matthewd :+1:

starbelly commented:
Ok, so unless `exit status == 0` then raise. Sounds good to me. Code coming up.

matthewd commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -228,14 +237,16 @@ def setup
       filename = "awesome file.sql"
-      Kernel.expects(:system).with("psql -q -f awesome\\ file.sql my-app-db")
+      args = [['psql', 'psql',], '-q', '-f', 'awesome\\ file.sql', 'my-app-db']
+      Kernel.expects(:system).with(*args).returns(true)
```
Why mocking is dangerous: a play in two lines.

(Yes, I know we don't have much choice here.)

starbelly replied:
Just realized that while that test passes, in the real world the code would fail. That is, can't use filenames with escaped spaces with the current implementation of run_cmd ( btw I don't like that name ).  Aside from performance, wouldn't it be feasible to re-write tests to use less mocking and test the real deal? You'll have to excuse me if I'm a bit naive on these issues.

matthewd replied:
It's not practical for these tests to actually run real commands, I don't think.

And yes, the problem is specific to your change: the space in the filename should not be escaped.

starbelly replied:
Hmmm yeah, bad.

matthewd commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -252,6 +259,19 @@ def load_seed
       end
 
+      def self.run_cmd(cmd, args, action)
+        fail "Invalid database command - #{cmd}" unless valid_bin? cmd
```
What's the purpose of this check?

starbelly replied:
I figured it would be best not run to accept arbitrary commands. Too much paranoia?

senny replied:
@starbelly I think so, let's remove that check.

seapy commented:
i like this feature +1
currently `structure_load` don't raise error, so people don't know what is wrong.

starbelly commented:
Right, so looking at this patch now without any hard code about command line programs, I'm thinking it should go somewhere else, since it's so general. I think run_cmd should accept accept a forth argument ( should be named params if so ), a block that will generate the error message. Also, the 'action' argument should be optional.

senny commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -252,6 +259,19 @@ def load_seed
 
+      def self.run_cmd(cmd, args, action)
+        fail "Invalid database command - #{cmd}" unless valid_bin? cmd
+        command = args.dup.unshift([cmd, cmd])
```
I'm not sure I understand why we need to `unshift` `[cmd, cmd]`. According to the docs, `Kernel.system(cmd, *args)` should do the same, no?

```
commandline                 : command line string which is passed to the standard shell
cmdname, arg1, ...          : command name and one or more arguments (no shell)
[cmdname, argv0], arg1, ... : command name, argv[0] and zero or more arguments (no shell)
```

http://ruby-doc.org/core-2.2.0/Kernel.html#method-i-system

senny commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -252,6 +251,18 @@ def load_seed
       end
 
+      def self.run_cmd(cmd, args, action)
+        command = args.dup.unshift([cmd, cmd])
```
I'm not sure I understand why we need to `unshift` `[cmd, cmd]`. According to the docs, `Kernel.system(cmd, *args)` should do the same, no?

```
commandline                 : command line string which is passed to the standard shell
cmdname, arg1, ...          : command name and one or more arguments (no shell)
[cmdname, argv0], arg1, ... : command name, argv[0] and zero or more arguments (no shell)
```

http://ruby-doc.org/core-2.2.0/Kernel.html#method-i-system

starbelly replied:
We don't. It only gives the process a specific name ("sometimes" according to docs). Will change it.

senny commented on activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:
```diff
@@ -48,18 +48,21 @@ def structure_dump(filename)
 
-        command = "pg_dump -i -s -x -O -f #{Shellwords.escape(filename)} #{search_path} #{Shellwords.escape(configuration['database'])}"
-        raise 'Error dumping database' unless Kernel.system(command)
+         args = ['-i', '-s', '-x', '-O', '-f', filename]
```
Indent looks off.

senny commented on activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:
```diff
@@ -48,18 +48,21 @@ def structure_dump(filename)
         search_path = configuration['schema_search_path']
         unless search_path.blank?
-          search_path = search_path.split(",").map{|search_path_part| "--schema=#{Shellwords.escape(search_path_part.strip)}" }.join(" ")
+          search_path = search_path.split(",").map{|search_path_part| "--schema=#{search_path_part.strip}" }.join(" ")
```
Can we move this block down and append directly to `args`?

senny commented:
@starbelly added some minor comments. Could you also add an entry to the CHANGELOG and push a rebased version?

starbelly commented:
Merged before rebasing and made a mess. Created a new PR, fixing this branch would be too messy. 
PR: https://github.com/rails/rails/pull/20632

senny commented:
@starbelly please reuse this PR. If you can create a new PR, you can also force-push that state onto this branch and Github will reflect that in this PR. This saves us from cutting the discussion in half.

starbelly commented:
@senny Sweet! Done.

senny commented on activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:
```diff
@@ -46,28 +46,21 @@ def purge
-          nil
-        when String
-          ActiveRecord::Base.dump_schemas
-        end
```
where did that case statement go?

starbelly commented:
Not my morning...

senny commented:
@starbelly no worries :), I'll have another look in a bit.

starbelly commented:
Cool, going to bed for now :+1:

senny commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -212,7 +211,7 @@ def load_schema_for(*args)
       end
 
-      def schema_file(format = ActiveRecord::Base.schema_format)
+      def schema_file(format = ActiveSupport::Base.schema_format)
```
This should continue to be `ActiveRecord::Base.schema_format`.

senny commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -36,7 +36,6 @@ class DatabaseNotSupported < StandardError; end # :nodoc:
     #   DatabaseTasks.create_current('production')
     module DatabaseTasks
       extend self
-
```
can we :scissors: this to keep the patch focused?

senny commented on activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:
```diff
@@ -46,7 +46,6 @@ def purge
 
       def structure_dump(filename)
         set_psql_env
-
```
can we :scissors: this to keep the patch focused?

senny commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -252,6 +251,17 @@ def load_seed
+        msg = "Error #{action} database: failed to execute "
+        msg << "`#{cmd} #{args.join(' ')}`. Please "
+        msg << "check that #{cmd} is present on this system, is in your PATH"
+        msg << ', and has proper permissions.'
```
Can we add some line breaks to the message. It currently looks like that:

```
$ bin/rake db:structure:dump
Error dumping database: failed to execute `pg_dump -i -s -x -O -f /Users/senny/Projects/playground/ruby/rails/rails-master/db/structure.sql test123_development`. Please check that pg_dump is present on this system, is in your PATH, and has proper permissions.
/Users/senny/Projects/rails/activerecord/lib/active_record/tasks/database_tasks.rb:256:in `run_cmd'
/Users/senny/Projects/rails/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:64:in `structure_dump'
/Users/senny/Projects/rails/activerecord/lib/active_record/tasks/database_tasks.rb:182:in `structure_dump'
/Users/senny/Projects/rails/activerecord/lib/active_record/railties/databases.rake:273:in `block (3 levels) in <top (required)>'
Tasks: TOP => db:structure:dump
(See full trace by running task with --trace)
```

Something along the lines of:

```
$ bin/rake db:structure:dump
rake aborted!
failed to execute:
pg_dump -i -s -x -O -f /Users/senny/Projects/playground/ruby/rails/rails-master/db/structure.sql test123_development

Please check that `pg_dump` is:
 - present on this system
 - in your PATH
 - has proper permissions.

/Users/senny/Projects/rails/activerecord/lib/active_record/tasks/database_tasks.rb:256:in `run_cmd'
/Users/senny/Projects/rails/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:64:in `structure_dump'
/Users/senny/Projects/rails/activerecord/lib/active_record/tasks/database_tasks.rb:182:in `structure_dump'
/Users/senny/Projects/rails/activerecord/lib/active_record/railties/databases.rake:273:in `block (3 levels) in <top (required)>'
Tasks: TOP => db:structure:dump
(See full trace by running task with --trace)
```

senny commented on activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:
```diff
@@ -55,19 +54,21 @@ def structure_dump(filename)
           ActiveRecord::Base.dump_schemas
         end
+        args = ['-i', '-s', '-x', '-O', '-f', filename]
         unless search_path.blank?
```
instead of a `unless blank?` we can formulate it positively: `if present?`.

senny commented on activerecord/lib/active_record/tasks/postgresql_database_tasks.rb:
```diff
@@ -55,19 +54,21 @@ def structure_dump(filename)
-        raise 'Error dumping database' unless Kernel.system(command)
-
+          args << search_path.split(',').map do |part|
+            "--schema=#{part.strip}"
```
Is it no longer necessary to `Shellwords.escape` the schema name?

senny commented:
Added some more comments. I also got some errors while running the tests. Might be worth double checking that everything is green:

```
ARCONN=postgresql ruby -Itest test/cases/tasks/postgresql_rake_test.rb

.................EE.EE

Finished in 0.099777s, 220.4917 runs/s, 180.4023 assertions/s.
22 runs, 18 assertions, 0 failures, 4 errors, 0 skips
```

starbelly commented:
@senny duly noted all your comments and will make adjustments soon.

starbelly commented:
On another note, if you're tired, go to sleep : )

starbelly commented:
@senny Successfully recovered from burn out : ) 
 Cleaned things up, postgresql task tests are passing now as well.

senny commented:
@starbelly thank you. I just checked out the latest code of this PR but still get [some errors](https://gist.github.com/senny/c7676928e1b9c0fbfc7a) when running the tests.

```
$ bin/test -a postgresql test/cases/tasks/postgresql_rake_test.rb
22 runs, 18 assertions, 0 failures, 4 errors, 0 skips
```

starbelly commented:
@senny I left this reply yesterday, but in the wrong place : 

Hmm, I just ran this test again (in rails-dev-box) : 

```
bin/test -a postgresql test/cases/tasks/postgresql_rake_test.rb
Using postgresql
Run options: -a postgresql --seed 64833

# Running:

......................

Finished in 0.127301s, 172.8193 runs/s, 172.8193 assertions/s.

22 runs, 22 assertions, 0 failures, 0 errors, 0 skips
```
 On my local machine I get the same results you did. However, if I use the same code outside of the test units, all works fine.

```
require './lib/active_record'
require './lib/active_record/tasks/database_tasks'
configuration = {
        'adapter'  => 'postgresql',
        'database' => 'my-app-db'
 }
db_config = YAML.load_file("./config.yml")
ActiveRecord::Base.default_timezone = :utc
ActiveRecord::Base.configurations = YAML.load_file("./config.yml")
ActiveRecord::Base.establish_connection(:development)
ActiveRecord::Tasks::DatabaseTasks.structure_dump(configuration, 'my-dump.sql')
```
 If I change the code in run_cmd to use backticks to capture the output, here's what we get: 

[archiver (db)] connection to database "my-app-db" failed: FATAL:  database "my-app-db" does not exist
pg_dump: [archiver (db)] connection to database "my-app-db" failed: FATAL:  database "my-app-db" does not exist
pg_dump: [archiver (db)] connection to database "my-app-db" failed: FATAL:  database "my-app-db" does not exist
pg_dump: [archiver (db)] connection to database "my-app-db" failed: FATAL:  database "my-app-db" does not exist
psql: FATAL:  database "my-app-db" does not exist
psql: FATAL:  database "my-app-db" does not exist

 This is also why I wanted check for error codes instead of just failing unless system returns true. However, someone had already stated something like this in another part of rails was done and turned into a mess, depending what platform you were on. Maybe a better solution is to simply execute the command in a named pipe and capture the output so we can try and give a more meaningful error.

 Please advise.

nateberkopec commented:
@senny @starbelly [Here's a git patch that makes this branch pass locally on my machine.](https://gist.github.com/nateberkopec/0c263d70835e86081e87).

starbelly commented:
@nateberkopec Danke! This works on my end as well. heh, after looking at your patch I immediately saw that I failed to commit a few lines of code. This is by far the sloppiest patch work I've ever done, especially for such a simple task. I have no good excuse :(
 
 Thanks once again

senny commented:
@starbelly can you push a rebased version?

starbelly commented:
@senny all done.

senny commented:
@starbelly when I look at the diff something doesn't feel quite right. Where did [this case statement](https://github.com/rails/rails/pull/17885/files#diff-9fa08d81184086c860434b970c3e2e80L50) and [the tests](https://github.com/rails/rails/pull/17885/files#diff-71dbe365b84aa1c378ee91d7d7b3ebb9L212) for it go?

```ruby
-        search_path = case ActiveRecord::Base.dump_schemas
-        when :schema_search_path
-          configuration['schema_search_path']
-        when :all
-          nil
-        when String
-          ActiveRecord::Base.dump_schemas
-        end
```

Commit:
Add run_cmd class method to ActiveRecord::Tasks::DatabaseTasks

  -   Added run_cmd() class method to dry up Kernel.system() messages within
      this namespace and avoid shell expansion by passing a list of
      arguments instead of a string

  -   Update structure_dump, structure_load, and related tests units to
      pass a list of params instead of using a string to
      avoid shell expansion

starbelly commented:
Hmmm, git rebase and I just are not getting along, hopefully soon we will.
 
@senny corrected.

senny commented:
I finally made some time to wrap this PR up.

I did change the error message to make it clear that the command can also fail due to other reasons. I put up [a gist](https://gist.github.com/senny/4e6eaa446ea2e3e696dc) with possible outputs. I also moved `run_cmd` into `PostgreSQLDatabaseTasks`. As it stands now the error will print the whole command. As we are passing username and password with environment variables this is fine for PG. Other adapters might pass the password using flags, which should not be logged. That's why I kept the method exclusive for the PG module for now.

@starbelly thanks for your work and sorry for the delay.

starbelly commented:
@senny Cool. Sorry that this PR was a mess, really shouldn't have been. Next time will be better :)
